### PR TITLE
add scripts/brewkit/python-venv.sh

### PR DESCRIPTION
**Pull request by magnusviri**
This file should be 755

```
[python-venv 2a1c711] add scripts/brewkit/python-venv.sh
 1 file changed, 42 insertions(+)
 create mode 100755 scripts/brewkit/python-venv.sh
```

### CODE REVIEW

File: scripts/brewkit/python-venv.sh [hunk 1]
```diff
@@ -0,0 +1,54 @@
+#!/bin/sh
+
+CMD_NAME=$1
+PROJECT_NAME=$2
+
+python -m venv $PREFIX/libexec
+
+cd "$PREFIX"
```
Comment (mxcl) [inline scripts/brewkit/python-venv.sh:8]: This should be a parameter to the script. I don’t think `$PREFIX` is set.

File: scripts/brewkit/python-venv.sh [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+sed -i.bak 's|VIRTUAL_ENV=".*"|VIRTUAL_ENV="$(cd "$(dirname "${BASH_SOURCE[0]}")"/.. \&\& pwd)"|' activate
+rm activate.bak
+
+# FIXME a lot: this "updates" the `venv` on each run for relocatability
+cat <<EOF>>activate
+
+sed -i.bak \\
+  -e "s|$TEA_PREFIX/python.org/v$PYTHON_VERSION|\$TEA_PREFIX/python.org/v$PYTHON_VERSION_MAJ|" \\
+  -e 's|bin/python$PYTHON_VERSION_MAJ.$PYTHON_VERSION_MIN|bin/python|' \\
+  -e "s|$PREFIX/libexec|\$TEA_PREFIX/$PROJECT_NAME/v$VERSION/libexec|" \\
```
Comment (mxcl) [inline scripts/brewkit/python-venv.sh:47]: `PYTHON_VERSION`, `VERSION` etc. are not set, so we will need to rethink this.

Comment (mxcl) [reply scripts/brewkit/python-venv.sh:47]: Actually this is broken anyway. I'll suggest a replacement.

In order to test this, make a small mod to `mesonbuild.com`’s package yml

**Comment by magnusviri** [comment]
This probably isn't the place to discuss this but I thought I'd post here to find out where I should move this.

Pulling python projects from https://pypi.org is probably better than from GitHub. One project that I need to create a package for is Ansible, and they don't even have their releases or tags posted there, they're here: https://pypi.org/project/ansible/. To do this, I'm guessing tea would need a way to get versions from there. And the URL would have to be looked up as well. For example, Ansible's latest version download URL is

https://files.pythonhosted.org/packages/18/fd/963a3328d6f72e54d17dfbb40af6e9d6827333a5f1745c79c6d81b30dc85/ansible-7.0.0.tar.gz

The previous version's URL is

https://files.pythonhosted.org/packages/06/75/6e1da76cbc323e5170ac9e63b7c86e17f4387614129ef2514b407c6080d2/ansible-6.6.0.tar.gz

So the URL is pretty unique for each version.

File: scripts/brewkit/python-venv.sh (hunk 2 above)
Comment (magnusviri) [reply scripts/brewkit/python-venv.sh:47]: I set them in the script. Line 18 https://github.com/teaxyz/pantry.extra/pull/45/files

**Comment by jonchang** [comment]
PyPI's preference is for downstream packagers to use the JSON API to look up the appropriate URL. You can use legacy URLs that will redirect to the right place, but setting something up to use the API is probably best (due to issues of e.g. yanked releases and so on)

```console
$ curl -I https://files.pythonhosted.org/packages/source/a/ansible/ansible-7.0.0.tar.gz

HTTP/2 302 
cache-control: max-age=604800, public
content-type: application/octet-stream
location: https://files.pythonhosted.org/packages/18/fd/963a3328d6f72e54d17dfbb40af6e9d6827333a5f1745c79c6d81b30dc85/ansible-7.0.0.tar.gz
```

File: scripts/brewkit/python-venv.sh (hunk 2 above)
Comment (mxcl) [reply scripts/brewkit/python-venv.sh:47]: we prefer args to variables generally.

**Comment by mxcl** [comment]
I think we should just rewrite the whole `pyvenv.cfg` rather than use `sed`.

You can do it with `PYTHON_PREFIX=$(which python)` `PYTHON_VERSION=$(python --version | cut -d' ' -f2)` and `$TEA_PREFIX` (which is set by `tea`)

**Comment by mxcl** [comment]
we do need to change meson to use this as the “test condition” for CI

File: scripts/brewkit/python-venv.sh (hunk 2 above)
Comment (magnusviri) [reply scripts/brewkit/python-venv.sh:47]: That's a total of 7 arguments. I'll make the switch though.

**Comment by magnusviri** [comment]
I see pip is using "--no-deps".

https://github.com/teaxyz/pantry.core/blob/c73c34df95cb251f4f60999f0aca16769da1173e/projects/mesonbuild.com/package.yml#L18

I tried building a package for https://github.com/kislyuk/yq and it built but didn't work because it requires several packages, argcomplete, toml, and xmltodict. I removed "--no-deps" to see if it would build with the dependencies but then it didn't even build.

How do we install the dependencies of a package?

Also, [here's the brew formula for yq](https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/python-yq.rb).

File: scripts/brewkit/python-venv.sh [hunk 3]
```diff
@@ -0,0 +1,59 @@
+#!/bin/sh
+
+CMD_NAME=$1
+PROJECT_NAME=$2
+VERSION=$3
+PREFIX=$4
+PYTHON_VERSION=$5
+PYTHON_VERSION_MAJ=$6
+PYTHON_VERSION_MIN=$7
```
Comment (mxcl) [inline scripts/brewkit/python-venv.sh:9]: Oof, well yeah variables is clearer. But I believe we can streamline this to just two vars, I'll do that after merge.

Still want to see meson build switched to using this so we can see it working though. You can do that in this PR.

Comment (magnusviri) [reply scripts/brewkit/python-venv.sh:9]: Ok, I added that.

**Comment by mxcl** [comment]
Great addition thanks @magnusviri

**Pull request merged**